misc/userscripts/add-nextcloud-bookmarks

```diff
@@ -0,0 +1,180 @@
+
+troubleshooting:
+    Errors detected within this userscript with have an exit of 231.  All other
+    exit codes will come from requests.
+"""
+
+import configparser
+from base64 import b64encode
+from os import environ, path
+from sys import argv, exit
+
+from PyQt5.QtWidgets import QApplication, QInputDialog, QLineEdit
+from requests import get, post
+
+
+def get_text(name, info):
+    """Get input from the user."""
+    _app = QApplication(argv)  # noqa: F841
+    if name == "password":
+        text, ok = QInputDialog.getText(
+            None,
+            "add-nextcloud-bookmarks userscript",
+            "Please enter {}".format(info),
+            QLineEdit.Password,
+        )
+    else:
+        text, ok = QInputDialog.getText(
+            None, "add-nextcloud-bookmarks userscript", "Please enter {}".format(info)
+        )
+    if not ok:
+        message("info", "Dialog box canceled.")
+        exit(0)
+    return text
+
+
+def message(level, text):
+    """display message"""
+    with open(environ["QUTE_FIFO"], "w") as fifo:
+        fifo.write(
+            'message-{} "add-nextcloud-bookmarks userscript: {}"'.format(level, text)
+        )
+        fifo.flush()
+
+
+if "QUTE_FIFO" not in environ:
+    print(
+        "This script is designed to run as a qutebrowser userscript, "
+        "not as a standalone script."
+    )
+    exit(231)
+
+if "QUTE_CONFIG_DIR" not in environ:
+    if "XDG_CONFIG_HOME" in environ:
+        QUTE_CONFIG_DIR = environ["XDG_CONFIG_HOME"] + "/qutebrowser"
+    else:
+        QUTE_CONFIG_DIR = environ["HOME"] + "/.config/qutebrowser"
+else:
+    QUTE_CONFIG_DIR = environ["QUTE_CONFIG_DIR"]
+
+config_file = QUTE_CONFIG_DIR + "/add-nextcloud-bookmarks.ini"
+if path.isfile(config_file):
+    config = configparser.ConfigParser()
+    config.read(config_file)
+    settings = dict(config.items("nextcloud"))
+else:
+    settings = {}
+
+settings_info = [
+    ("host", "host information.", "required"),
+    ("user", "username.", "required"),
+    ("password", "password.", "required"),
+    ("description", "description or leave blank", "optional"),
+    ("tags", "tags (comma separated) or leave blank", "optional"),
+]
+
+# check for settings that need user interaction and clear optional setting if need be
+for setting in settings_info:
+    if setting[0] not in settings:
+        userInput = get_text(setting[0], setting[1])
+        settings[setting[0]] = userInput
+    if setting[2] == "optional":
+        if settings[setting[0]] == "None":
+            settings[setting[0]] = ""
+
+# tags need to be formatted like [], ["one"] or ["one", two"] as a string
+tags = settings["tags"]
+settings["tags"] = str(tags.split(", ")).replace("'", '"')
+
+QUTE_URL = environ["QUTE_URL"]
+api_url = settings["host"] + "/index.php/apps/bookmarks/public/rest/v2/bookmark"
+
+auth = b64encode(format(settings["user"] + ":" + settings["password"]).encode("utf-8"))
```

Comment:
you can just pass an auth=(username, password) kwarg to requests and it'll take care of this for you

Reply:
While requests supports this, Nextcloud Apps requires [encoding][0].  I get 500 status codes without the encoding.  


[0]: https://nextcloudappstore.readthedocs.io/en/latest/restapi.html#authentication

Reply:
requests definitely does proper basic auth: https://requests.readthedocs.io/en/master/user/authentication/

Reply:
I agree this should work properly without having to do this by hand. What exactly did you call when you get the 500 error from nextcloud?

Reply:
Ah, I see you did that already.